### PR TITLE
Base64-encode userID with invalid chars

### PR DESCRIPTION
**Pull request by sbryzak**
This PR modifies the UserAccount controller to check whether the `UserAccount.Spec.UserID` property contains any invalid characters, and if so performs a base64-encoding on the UserID value and uses that instead when creating the `Identity`.

E2e tests: https://github.com/codeready-toolchain/toolchain-e2e/pull/444

Fixes https://issues.redhat.com/projects/CRT/issues/CRT-1410

### CODE REVIEW

**Comment by sbryzak** [comment]
Tests still being worked on, will be added soon.

File: controllers/useraccount/useraccount_controller.go [hunk 1]
```diff
@@ -325,8 +344,8 @@ func (r *Reconciler) ensureIdentity(logger logr.Logger, config membercfg.Configu
 	if userAcc.Spec.OriginalSub != "" {
 
 		// Encoded the OriginalSub as an unpadded Base64 value
-		encodedName := fmt.Sprintf("b64:%s", base64.RawStdEncoding.EncodeToString([]byte(userAcc.Spec.OriginalSub)))
-		_, createdOrUpdated, err := r.loadIdentityAndEnsureMapping(logger, config, encodedName, userAcc, user)
+		//encodedName := fmt.Sprintf("b64:%s", base64.RawStdEncoding.EncodeToString([]byte(userAcc.Spec.OriginalSub)))
```
Comment (alexeykazakov) [inline controllers/useraccount/useraccount_controller.go:347]:
```suggestion
```

File: controllers/useraccount/useraccount_controller.go [hunk 2]
```diff
@@ -811,3 +840,16 @@ func (r *Reconciler) lookupAndDeleteCheUser(logger logr.Logger, config membercfg
...
+func isIdentityNameCompliant(name string) bool {
+	if len(name) > 253 {
+		return false
+	}
+
+	if !dns1123ValueRegexp.MatchString(name) {
+		return false
+	}
+
+	return true
```
Comment (alexeykazakov) [inline controllers/useraccount/useraccount_controller.go:854]:
```suggestion
	return dns1123ValueRegexp.MatchString(name)
```

File: controllers/useraccount/useraccount_controller.go [hunk 3]
```diff
@@ -811,3 +840,16 @@ func (r *Reconciler) lookupAndDeleteCheUser(logger logr.Logger, config membercfg
 
 	return nil
 }
+
+// isIdentityNameCompliant returns true if the specified name is RFC-1123 compliant, otherwise it returns false
+func isIdentityNameCompliant(name string) bool {
+	if len(name) > 253 {
+		return false
```
Comment (alexeykazakov) [inline controllers/useraccount/useraccount_controller.go:847]: Please add a unit test for that.

File: controllers/useraccount/useraccount_controller_test.go [hunk 4]
```diff
@@ -1373,6 +1373,61 @@ func TestReconcile(t *testing.T) {
...
+		t.Run("create user identity mapping", func(t *testing.T) {
+			r, req, _, _ := prepareReconcile(t, username, userAcc, preexistingUser)
+			//when
+			res, err := r.Reconcile(context.TODO(), req)
+
+			//then
+			require.NoError(t, err)
+			assert.Equal(t, reconcile.Result{}, res)
+
+			updatedUser := assertUser(t, r, userAcc)
```
Comment (alexeykazakov) [inline controllers/useraccount/useraccount_controller_test.go:1391]: Let's add a comment here that only a User is expected to be created in the first reconcile.

File: controllers/useraccount/useraccount_controller_test.go [hunk 5]
```diff
@@ -1373,6 +1373,61 @@ func TestReconcile(t *testing.T) {
...
+					toolchainv1alpha1.Condition{
+						Type:   toolchainv1alpha1.ConditionReady,
+						Status: corev1.ConditionFalse,
+						Reason: "Provisioning",
+					})
+
+				// Check the created/updated identity
+				identity1 := assertIdentity(t, r, userAcc, config.Auth().Idp())
+
+				t.Run("reconcile once more to ensure the users", func(t *testing.T) {
```
Comment (alexeykazakov) [inline controllers/useraccount/useraccount_controller_test.go:1416]:
```suggestion
				t.Run("reconcile once more to ensure the user-identity mapping", func(t *testing.T) {
```

**Review by alexeykazakov — APPROVED** [review]
Looks good but please take a look at my minor comments.

File: controllers/useraccount/useraccount_controller.go (hunk 3 above)
Comment (sbryzak) [reply controllers/useraccount/useraccount_controller.go:846]: Done

File: controllers/useraccount/useraccount_controller_test.go (hunk 4 above)
Comment (sbryzak) [reply controllers/useraccount/useraccount_controller_test.go:1417]: Done

File: controllers/useraccount/useraccount_controller_test.go (hunk 5 above)
Comment (sbryzak) [reply controllers/useraccount/useraccount_controller_test.go]: Done

File: controllers/useraccount/useraccount_controller.go (hunk 2 above)
Comment (sbryzak) [reply controllers/useraccount/useraccount_controller.go]: Fixed

File: controllers/useraccount/useraccount_controller.go (hunk 1 above)
Comment (sbryzak) [reply controllers/useraccount/useraccount_controller.go]: Fixed

**Review by alexeykazakov — APPROVED** [review]

**Pull request merged**